Swabbr.Api/Authentication/TokenService.cs

```diff
@@ -1,65 +1,179 @@
-                signingCredentials: credentials
+                notBefore: DateTime.UtcNow,
+                expires: DateTime.UtcNow.AddMinutes(Convert.ToDouble(_options.TokenValidity)),
+                signingCredentials: signingCredentials
             );
 
+            // Generate and store a new refresh token.
+            var refreshToken = GenerateRefreshToken();
+            await _userRefreshTokenRepository.StoreRefreshTokenHashAsync(userId, refreshToken);
+
             return new TokenWrapper
             {
-                UserId = user.Id,
+                UserId = userId,
                 DateCreated = DateTimeOffset.Now,
                 Token = new JwtSecurityTokenHandler().WriteToken(token),
-                TokenExpirationTimespan = TimeSpan.FromMinutes(_options.TokenValidity)
+                TokenExpirationInMinutes = _options.TokenValidity, // Options is in minutes.
+                RefreshToken = refreshToken,
+                RefreshTokenExpirationInMinutes = _options.RefreshTokenValidity, // Options is in minutes.
+            };
+        }
+
+        /// <summary>
+        ///     Refresh an expired authentication token and generate a new <see cref="TokenWrapper"/>
+        ///     containing a new refresh token as well. The expiration for the refresh token is checked
+        ///     based on the valid-from date of the token and the <see cref="_options"/> values.
+        /// </summary>
+        /// <param name="expiredToken">The expired jwt token.</param>
+        /// <param name="refreshToken">Valid refresh token.</param>
+        /// <returns>New token wrapper with new refresh token.</returns>
+        internal virtual async Task<TokenWrapper> RefreshTokenAsync(string expiredToken, string refreshToken)
+        {
+            // First extract the user id. This method will throw if the expiredToken has an invalid format.
+            var principal = GetPrincipalFromExpiredToken(expiredToken);
+            var identity = principal.Identity as ClaimsIdentity; // Cast required to access the claims.
+            if (!Guid.TryParse(identity?.FindFirst(ClaimTypes.NameIdentifier)?.Value, out Guid userId))
+            {
+                throw new AuthenticationException("Couldn't extract user id from token principal");
+            }
+
+            // Validate refresh token in db based on user id
+            if (!await _userRefreshTokenRepository.ValidateRefreshTokenAsync(userId, refreshToken))
+            {
+                throw new AuthenticationException("Refresh token was invalid");
+            }
+
+            // Then validate the lifetime.
+            var tokenHandler = new JwtSecurityTokenHandler();
+            var jwtToken = tokenHandler.ReadJwtToken(expiredToken);
+            if (DateTimeOffset.UtcNow >= jwtToken.ValidFrom.AddMinutes(_options.RefreshTokenValidity))
+            {
+                throw new AuthenticationException("Refresh token has expired");
+            }
```

Comment:
You could validate the lifetime before validating against the database. This avoids unnecessary database queries if the token turns out to be expired anyway.

Reply:
Good point, implemented